### PR TITLE
Blocks: Implement insert_hooked_blocks()

### PR DESCRIPTION
**Pull request by ockham**
WIP. See [`#59313`](https://core.trac.wordpress.org/ticket/59313) and https://github.com/WordPress/wordpress-develop/pull/5158.

For [`#59313`](https://core.trac.wordpress.org/ticket/59313), we need a function that will plug into parsed block tree traversal and serialization (https://github.com/WordPress/wordpress-develop/pull/5246). For each block traversed, it will obtain the list of all hooked blocks that use it as their anchor block (`get_hooked_blocks`, https://github.com/WordPress/wordpress-develop/pull/5239), and insert those hooked blocks into the parsed block tree (`insert_inner_block`, `prepend_inner_block`, `append_inner_block`, https://github.com/WordPress/wordpress-develop/pull/5240).

TODO:
- [ ] Add test coverage for all four possible relative insertion positions.

---

For early testing, use the [Like Button block](https://github.com/ockham/like-button/releases/download/v0.4.1/like-button.zip) and apply the following diff:

```diff
diff --git a/src/wp-includes/class-wp-block-patterns-registry.php b/src/wp-includes/class-wp-block-patterns-registry.php
index a83b35185b..7f617dbbea 100644
--- a/src/wp-includes/class-wp-block-patterns-registry.php
+++ b/src/wp-includes/class-wp-block-patterns-registry.php
@@ -165,7 +165,10 @@ final class WP_Block_Patterns_Registry {
                        return null;
                }
 
-               return $this->registered_patterns[ $pattern_name ];
+               $pattern            = $this->registered_patterns[ $pattern_name ];
+               $blocks             = parse_blocks( $pattern['content'] );
+               $pattern['content'] = traverse_and_serialize_blocks( $blocks, 'insert_hooked_blocks' );
+               return $pattern;
        }
 
        /**
```

Trac ticket: https://core.trac.wordpress.org/ticket/59399

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/blocks.php [hunk 1]
```diff
@@ -760,6 +760,29 @@ function get_hooked_blocks( $name ) {
...
+			'blockName'    => $hooked_block_type,
+			'attrs'        => array(),
+			'innerHTML'    => '',
+			'innerContent' => array(),
+			'innerBlocks'  => array(),
+		);
+		if ( 'before' === $relative_position ) {
+			insert_inner_block( $parent, $block_index, $chunk_index, $hooked_block );
+		} elseif ( 'after' === $relative_position ) {
+			insert_inner_block( $parent, $block_index + 1, $chunk_index, $hooked_block );
```
Comment (ockham) [inline src/wp-includes/blocks.php:776]: Add checks to make sure `$parent` (and `$block_index` and `$chunk_index`) exist -- 'cause they don't [if `$block` is at the top level of the tree](https://github.com/WordPress/wordpress-develop/pull/5246/files#diff-8c99af92e4ec0fdb307ddd9b42be1e1ef1efe4a9f31287c23f346244dddd1ce9R954-R955) 😬

File: src/wp-includes/blocks.php [hunk 2]
```diff
@@ -760,6 +760,29 @@ function get_hooked_blocks( $name ) {
...
+		} elseif ( 'first_child' === $relative_position ) {
+			prepend_inner_block( $block, $hooked_block );
```
Comment (gziolo) [inline src/wp-includes/blocks.php:778]: One observation. In this scenario, I assume that `$parent` is the same as `$block`. The code would be easier to follow if that was explicitly expresses as `$parent`. 

By the way, for consistency, the variable could become `$parent_block`.

File: src/wp-includes/blocks.php [hunk 3]
```diff
@@ -760,6 +760,29 @@ function get_hooked_blocks( $name ) {
...
+	foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
+		$hooked_block = array(
+			'blockName'    => $hooked_block_type,
+			'attrs'        => array(),
+			'innerHTML'    => '',
+			'innerContent' => array(),
+			'innerBlocks'  => array(),
+		);
+		if ( 'before' === $relative_position ) {
+			insert_inner_block( $parent, $block_index, $chunk_index, $hooked_block );
```
Comment (ockham) [inline src/wp-includes/blocks.php:774]: I had a realization last night that `before` (and probably `after`) insertion isn't going to work like this in combination with tree traversal and serialization 😕 The reason is that we're iterating over `parent`'s `innerBlocks` array in `traverse_and_serialize_blocks` when we encounter this line, which then attempts to modify that array (by inserting another element before or after the current one, respectively).

I came up with an idea how to fix that. The good news is that if it works, we'll be able to get rid of our `$chunk_index` arithmetics, and even remove `insert_inner_block` and friends. The bad news is that I'll have to change the function signature of the callback in `traverse_and_serialize_blocks` once again 😬

File: src/wp-includes/blocks.php (hunk 2 above)
Comment (ockham) [reply src/wp-includes/blocks.php:778]: > One observation. In this scenario, I assume that `$parent` is the same as `$block`.

They're different, see https://github.com/WordPress/wordpress-develop/pull/5247/files#diff-8c99af92e4ec0fdb307ddd9b42be1e1ef1efe4a9f31287c23f346244dddd1ce9R763.

`$block` is the block that tree traversal is currently processing, `$parent` is its parent. First child/last child insertion thus needs to happen to `$block` (whereas sibling insertion needs to happen to `$parent`'s `innerBlocks`; however, note my [other comment](https://github.com/WordPress/wordpress-develop/pull/5247/files#r1331151853)).

**Comment by gziolo** [comment]
I assume we no longer need it after landing an alternative version. I see now that https://core.trac.wordpress.org/ticket/59399 is closed, so I will close this one, too.

**Pull request closed without merging**